text/0000-builder-spec.md

```diff
@@ -0,0 +1,282 @@
+* `trust-builder <builder>` &rarr; Trust builder
+* `untrust-builder <builder>` &rarr; Stop trusting builder
+* `list-trusted-builders` &rarr; List Trusted Builders
+* `create-builder` &rarr; Create builder image
+
+Users of `pack-orb` and `tekton` interact with Builders through supplying a builder to use for building applications. Users of `kpack` can create and use builders.
+
+A sample inspected builder (using `pack inspect-builder`), looks like:
+```bash
+ $  pack inspect-builder cnbs/sample-builder:bionic
+Inspecting builder: cnbs/sample-builder:bionic
+
+REMOTE:
+
+Created By:
+  Name: Pack CLI
+  Version: 0.13.1+git-4134cc6.build-1135
+
+Trusted: No
+
+Stack:
+  ID: io.buildpacks.samples.stacks.bionic
+
+Lifecycle:
+  Version: 0.9.1
+  Buildpack APIs:
+    Deprecated: (none)
+    Supported: 0.2, 0.3, 0.4
+  Platform APIs:
+    Deprecated: (none)
+    Supported: 0.3, 0.4
+
+Run Images:
+  cnbs/sample-stack-run:bionic
+
+Buildpacks:
+  ID                            VERSION        HOMEPAGE
+  samples/java-maven            0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/java-maven
+  samples/kotlin-gradle         0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/kotlin-gradle
+  samples/ruby-bundler          0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/ruby-bundler
+  samples/hello-universe        0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/hello-universe
+  samples/hello-world           0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/hello-world
+  samples/hello-moon            0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/hello-moon
+
+Detection Order:
+ ├ Group #1:
+ │  └ samples/java-maven@0.0.1
+ ├ Group #2:
+ │  └ samples/kotlin-gradle@0.0.1
+ ├ Group #3:
+ │  └ samples/ruby-bundler@0.0.1
+ └ Group #4:
+    └ samples/hello-universe@0.0.1
+       └ Group #1:
+          ├ samples/hello-world@0.0.1
+          └ samples/hello-moon@0.0.1
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+## Structure of Builders
+### Files/Directories
+A builder is composed of at least the following directories/files:
+```
+- /cnb/buildpacks/...<buildpack ID>/<buildpack version>/
+- /cnb/lifecycle/<lifecycle binaries>
+- /cnb/order.toml
+- /cnb/stack.toml
+```
+
+### Env Vars/Labels
+A builder's environment is the build-time environment of the stack, and as such, requires all of the [build image specifications][build-image-specs].
+
+Additionally, a builder requires:
+* The image config's WorkingDir should be set
+* The image config's Env field has the environment variable `CNB_APP_DIR` set to the application directory of the build environment
+* The image config's Env field has the environment variable `CNB_LAYERS_DIR` set
+* The image config's Env field has the environment variable `CNB_PLATFORM_DIR` set
+* The image config's Env field has the environment variable `CNB_PLATFORM_DIR` set
```

Comment:
We should specify that the directories these env vars point to should exist in the builder image.

Reply:
Fixed in 94207149e7534c608c8b926652ba12cdccff66b8